File: src/lucida/models/view_state.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


AxisSelectorKind = Literal["index", "range", "set"]
RenderMode = Literal["2d", "3d"]


class ModelBase(BaseModel):
    model_config = ConfigDict(extra="forbid")


class DatasetRef(ModelBase):
    dataset_id: str = Field(min_length=1)
    multiscale_name: str = Field(min_length=1)


class Viewport(ModelBase):
    width_px: int = Field(ge=1)
    height_px: int = Field(ge=1)
    pixel_ratio: float = Field(default=1.0, ge=0.5)


class AxisSelector(ModelBase):
    axis: str = Field(min_length=1)
    kind: AxisSelectorKind
    index: int | None = Field(default=None, ge=0)
    start: int | None = Field(default=None, ge=0)
    end_exclusive: int | None = Field(default=None, ge=1)
    indices: list[int] | None = None
    clamp: bool = True

    @model_validator(mode="after")
    def validate_selector_shape(self) -> "AxisSelector":
        if self.kind == "index":
            if self.index is None:
                raise ValueError("index selector requires index.")
            return self
        if self.kind == "range":
            if self.start is None or self.end_exclusive is None:
                raise ValueError("range selector requires start and end_exclusive.")
            return self
        if self.kind == "set":
            if self.indices is None:
                raise ValueError("set selector requires indices.")
            return self
        return self


class SlabSettings(ModelBase):
    thickness_vox: int = Field(default=1, ge=1)
    mode: Literal["single", "mip", "mean"] = "single"


class SliceSettings(ModelBase):
    axis: str | None = None
    index: int | None = Field(default=None, ge=0)
    slab: SlabSettings | None = None


class Camera2D(ModelBase):
    center_world: tuple[float, float]
    zoom: float = Field(ge=0.000001)
    rotation_deg: float = 0.0


class View2D(ModelBase):
    plane: Literal["xy", "xz", "yz"] = "xy"
    slice: SliceSettings | None = None
    camera: Camera2D


class LayerSource(ModelBase):
    multiscale_name: str | None = None
    array_path: str | None = None


class ChannelContrast(ModelBase):
    policy: Literal["fixed", "percentile"] = "percentile"
    min: float | None = None
    max: float | None = None
    p_low: float = 1.0
    p_high: float = 99.0


class ImageChannelSettings(ModelBase):
    index: int = Field(ge=0)
    enabled: bool
    color_rgba: tuple[float, float, float, float] | None = None
    contrast: ChannelContrast | None = None
    gamma: float = Field(default=1.0, ge=0.01)


class ImageLayerSettings(ModelBase):
    channel_mode: Literal["single", "rgb", "composite"] = "composite"
    channels: list[ImageChannelSettings] = Field(default_factory=list)
    interpolation: Literal["nearest", "linear"] = "linear"


class LabelLayerSettings(ModelBase):
    outline: bool = True
    outline_width_px: int = Field(default=1, ge=0)
    show_fill: bool = True


class LayerState(ModelBase):
    layer_id: str = Field(min_length=1)
    type: Literal["image", "labels", "annotations"]
    dataset_id: str | None = None
    source: LayerSource | None = None
    visible: bool
    opacity: float = Field(ge=0, le=1)
    image: ImageLayerSettings | None = None
    labels: LabelLayerSettings | None = None


class RenderSettings(ModelBase):
    background_rgba: tuple[float, float, float, float] | None = None


class PerformanceHints(ModelBase):
    quality: Literal["draft", "final"] = "draft"
    target_frame_ms: int = Field(default=200, ge=1)
    progressive: bool = True
    lod_mode: Literal["auto", "fixed"] = "auto"
    fixed_level: int | None = Field(default=None, ge=0)
    max_cpu_cache_bytes: int | None = Field(default=None, ge=0)
    max_gpu_cache_bytes: int | None = Field(default=None, ge=0)
    prefer_gpu: bool = True


class ViewState(ModelBase):
    schema_version: Literal[1] = 1
    view_id: str = Field(min_length=1)
    session_id: str = Field(min_length=1)
    created_at: datetime | None = None
    mode: RenderMode
    datasets: list[DatasetRef] = Field(min_length=1)
    viewport: Viewport
    selectors: list[AxisSelector] = Field(min_length=1)
    view_2d: View2D | None = None
    view_3d: dict[str, Any] | None = None
    layers: list[LayerState] = Field(min_length=1)
    render_settings: RenderSettings | None = None
    performance: PerformanceHints | None = None
    state_hash: str | None = None
    state_version: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def validate_mode_dependencies(self) -> "ViewState":
        if self.mode == "2d" and self.view_2d is None:
            raise ValueError("mode=2d requires view_2d.")
        if self.mode == "3d" and self.view_3d is None:
            raise ValueError("mode=3d requires view_3d.")
        return self

File: crates/lucida-daemon/src/dto/view_state.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

fn default_schema_version() -> u8 {
    1
}

fn default_true() -> bool {
    true
}

fn default_pixel_ratio() -> f64 {
    1.0
}

fn default_thickness_vox() -> u64 {
    1
}

fn default_gamma() -> f64 {
    1.0
}

fn default_p_low() -> f64 {
    1.0
}

fn default_p_high() -> f64 {
    99.0
}

fn default_target_frame_ms() -> u64 {
    200
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum AxisSelectorKind {
    Index,
    Range,
    Set,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum RenderMode {
    #[serde(rename = "2d")]
    TwoD,
    #[serde(rename = "3d")]
    ThreeD,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct DatasetRef {
    pub dataset_id: String,
    pub multiscale_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct Viewport {
    pub width_px: u64,
    pub height_px: u64,
    #[serde(default = "default_pixel_ratio")]
    pub pixel_ratio: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct AxisSelector {
    pub axis: String,
    pub kind: AxisSelectorKind,
    pub index: Option<i64>,
    pub start: Option<i64>,
    pub end_exclusive: Option<i64>,
    pub indices: Option<Vec<i64>>,
    #[serde(default = "default_true")]
    pub clamp: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum SlabMode {
    Single,
    Mip,
    Mean,
}

impl Default for SlabMode {
    fn default() -> Self {
        Self::Single
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SlabSettings {
    #[serde(default = "default_thickness_vox")]
    pub thickness_vox: u64,
    #[serde(default)]
    pub mode: SlabMode,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct SliceSettings {
    pub axis: Option<String>,
    pub index: Option<i64>,
    pub slab: Option<SlabSettings>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct Camera2D {
    pub center_world: (f64, f64),
    pub zoom: f64,
    pub rotation_deg: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum Plane2D {
    Xy,
    Xz,
    Yz,
}

impl Default for Plane2D {
    fn default() -> Self {
        Self::Xy
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct View2D {
    #[serde(default)]
    pub plane: Plane2D,
    pub slice: Option<SliceSettings>,
    pub camera: Camera2D,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct LayerSource {
    pub multiscale_name: Option<String>,
    pub array_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ChannelContrastPolicy {
    Fixed,
    Percentile,
}

impl Default for ChannelContrastPolicy {
    fn default() -> Self {
        Self::Percentile
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct ChannelContrast {
    #[serde(default)]
    pub policy: ChannelContrastPolicy,
    pub min: Option<f64>,
    pub max: Option<f64>,
    #[serde(default = "default_p_low")]
    pub p_low: f64,
    #[serde(default = "default_p_high")]
    pub p_high: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct ImageChannelSettings {
    pub index: u64,
    pub enabled: bool,
    pub color_rgba: Option<[f64; 4]>,
    pub contrast: Option<ChannelContrast>,
    #[serde(default = "default_gamma")]
    pub gamma: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ChannelMode {
    Single,
    Rgb,
    Composite,
}

impl Default for ChannelMode {
    fn default() -> Self {
        Self::Composite
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum InterpolationMode {
    Nearest,
    Linear,
}

impl Default for InterpolationMode {
    fn default() -> Self {
        Self::Linear
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct ImageLayerSettings {
    #[serde(default)]
    pub channel_mode: ChannelMode,
    pub channels: Vec<ImageChannelSettings>,
    #[serde(default)]
    pub interpolation: InterpolationMode,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct LabelLayerSettings {
    #[serde(default = "default_true")]
    pub outline: bool,
    #[serde(default = "default_thickness_vox")]
    pub outline_width_px: u64,
    #[serde(default = "default_true")]
    pub show_fill: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum LayerType {
    Image,
    Labels,
    Annotations,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct LayerState {
    pub layer_id: String,
    #[serde(rename = "type")]
    pub layer_type: LayerType,
    pub dataset_id: Option<String>,
    pub source: Option<LayerSource>,
    pub visible: bool,
    pub opacity: f64,
    pub image: Option<ImageLayerSettings>,
    pub labels: Option<LabelLayerSettings>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct RenderSettings {
    pub background_rgba: Option<[f64; 4]>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RenderQuality {
    Draft,
    Final,
}

impl Default for RenderQuality {
    fn default() -> Self {
        Self::Draft
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum LodMode {
    Auto,
    Fixed,
}

impl Default for LodMode {
    fn default() -> Self {
        Self::Auto
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct PerformanceHints {
    #[serde(default)]
    pub quality: RenderQuality,
    #[serde(default = "default_target_frame_ms")]
    pub target_frame_ms: u64,
    #[serde(default = "default_true")]
    pub progressive: bool,
    #[serde(default)]
    pub lod_mode: LodMode,
    pub fixed_level: Option<u64>,
    pub max_cpu_cache_bytes: Option<u64>,
    pub max_gpu_cache_bytes: Option<u64>,
    #[serde(default = "default_true")]
    pub prefer_gpu: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct ViewState {
    #[serde(default = "default_schema_version")]
    pub schema_version: u8,
    pub view_id: String,
    pub session_id: String,
    pub created_at: Option<DateTime<Utc>>,
    pub mode: RenderMode,
    pub datasets: Vec<DatasetRef>,
    pub viewport: Viewport,
    pub selectors: Vec<AxisSelector>,
    pub view_2d: Option<View2D>,
    pub view_3d: Option<Value>,
    pub layers: Vec<LayerState>,
    pub render_settings: Option<RenderSettings>,
    pub performance: Option<PerformanceHints>,
    pub state_hash: Option<String>,
    #[serde(default)]
    pub state_version: u64,
}
